sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala

```diff
@@ -148,6 +149,10 @@ object AnsiTypeCoercion extends TypeCoercionBase {
       // interval type the string should be promoted as. There are many possible interval
       // types, such as year interval, month interval, day interval, hour interval, etc.
       case (_: StringType, _: AnsiIntervalType) => None
+      // [SPARK-50060] If a binary operation contains at least one collated string types, we can't
+      // decide which collation the result should have.
+      case (d1: StringType, d2: StringType) if SchemaUtils.hasNonUTF8BinaryCollation(d1) ||
```

Comment:
By design we should not hit this rule. The CollationTypeCast is done first in the TypeCoercion and AnsiTypeCoercion rules. @cloud-fan am I missing something here?

Reply:
`CollationTypeCast` will work for expressions, but it will not work for set operators such as `union`. However, this can be removed since we do not use it when we have collated string column in the output